README.md
# Dart Angles

[Official Dart package](https://pub.dartlang.org/packages/angles)

This package provides an `Angle` type seeking to eliminate the use of variable suffixes like `...deg` or `...rad`
as well as manual angle unit conversions such as `x * rad2deg` and the likes.

Therefore, this package's *raison d'être* is not to serve features but to increase code maintainability,
thus reducing maintaince cost and bugs.

Angles can be constructed from these units: degrees, radians, gradians and turns.

| Unit | Quantity of a full turn |
|--|--|
| Degree | 360° |
| Radian | 2π |
| Gradian | 400ᵍ |
| Turn | 1 |


## Sample code


```dart
// Various ways to define a half turn:
a0 = Angle.degrees(180.0);
a1 = Angle.radians(math.pi);
a2 = Angle.turns(0.5);

a0.degrees; // = 180.0
a1.radians; // = 3.1415...
a2.turns;   // = 0.5
```


## Documentation


Constructors:
- `Angle.degrees(x)`
- `Angle.radians(x)`
- `Angle.gradians(x)`
- `Angle.turns(x)`

Accessors:
- `.degrees()`
- `.radians()`
- `.gradians()`
- `.turns()`

Trigonometric functions:
- `.sin()`
- `.cos()`
- `.tan()`
- `Angle.asin(x)`
- `Angle.acos(x)`
- `Angle.atan(x)`
- `Angle.atan2(y, x)`
- `Angle.atanFullTurn(y, x)`
- `Angle.cartesian(x, y)`

Utility functions:
- `.toString()`
- `.approximately(other, range)`
- `.ratio(other)`
- `.normalized`

All arithmetic and relational operators are defined on angles as well.


CHANGELOG.md
# Changelog

## 2.1.1

- Change test import to absolute path: https://github.com/Jim-Eckerlein/dart-angles/pull/5

## 2.1.0

- Add `normalized` accessor
- Add `Angle.cartesian` constructor

## 2.0.0

- Updated SDK version.
- Add a `ratio()` function.
- `Angle.atanFullTurn()` now creates angles which actually make sense.
- `range` argument of `approximately()` must be non-negative.

## 1.2.0

- Add `.approximately()`

## 1.0.0

- Initial version


test/angles_test.dart
import 'package:test/test.dart';

import 'package:angles/angles.dart';

void main() {
  group("Arithmetical operations", () {
    test("Zero angle", () {
      expect(Angle.zero() + Angle.turns(1), Angle.turns(1));
      expect(Angle.turns(1) * 0, Angle.zero());
    });

    test("Multiplication", () {
      expect(Angle.turns(1) * 1, Angle.turns(1));
      expect(Angle.turns(1) * 2, Angle.turns(2));
      expect(-Angle.turns(2), Angle.turns(-2));
    });

    test("Division", () {
      expect(Angle.turns(5) / 1, Angle.turns(5));
      expect(Angle.turns(6) / 3, Angle.turns(2));
    });

    test("Ratio", () {
      expect(Angle.turns(3).ratio(Angle.turns(3)), 1);
      expect(Angle.turns(6).ratio(Angle.turns(3)), 2);
    });

    test("Addition", () {
      expect(Angle.turns(2) + Angle.turns(3), Angle.turns(5));
      expect(Angle.turns(2) - Angle.turns(3), Angle.turns(-1));
    });
  });
  group("Relational operators", () {
    test("Inequality", () {
      expect(Angle.turns(1).compareTo(Angle.turns(2)), -1);
      expect(Angle.turns(2).compareTo(Angle.turns(1)), 1);
      expect(Angle.turns(1) != Angle.turns(2), true);
      expect(Angle.turns(1) < Angle.turns(2), true);
      expect(Angle.turns(1) <= Angle.turns(2), true);
      expect(Angle.turns(2) > Angle.turns(1), true);
      expect(Angle.turns(2) >= Angle.turns(1), true);
    });
    test("Equality", () {
      expect(Angle.turns(2).compareTo(Angle.turns(2)), 0);
      expect(Angle.turns(2) <= Angle.turns(2), true);
      expect(Angle.turns(2) >= Angle.turns(2), true);
      expect(Angle.turns(2) == Angle.turns(2), true);
    });
  });
  group("Units", () {
    test("Turns", () {
      expect(Angle.turns(0), Angle.zero());
      expect(Angle.turns(0.5), Angle.halfTurn());
      expect(Angle.turns(0.25), Angle.quarterTurn());
      expect(Angle.turns(0.125), Angle.eighthTurn());
      expect(Angle.turns(1), Angle.fullTurn());
      expect(Angle.turns(1 / 3), Angle.thirdTurn());
    });
    test("Degrees", () {
      expect(Angle.degrees(0), Angle.zero());
      expect(Angle.degrees(90), Angle.quarterTurn());
      expect(Angle.degrees(360), Angle.fullTurn());
    });
    test("Gradians", () {
      expect(Angle.gradians(0), Angle.zero());
      expect(Angle.gradians(100), Angle.quarterTurn());
      expect(Angle.gradians(400), Angle.fullTurn());
    });
  });
  group("Utility functions", () {
    test("Stringification", () {
      expect(Angle.degrees(45).toString(), "45.0°");
    });
    test("Apprixmate equality", () {
      final Angle x = Angle.degrees(45);
      expect(x.approximately(Angle.degrees(45.01), 0.1), true);
      expect(x.approximately(Angle.degrees(44.99), 0.1), true);
      expect(x.approximately(Angle.degrees(45.1), 0.1), true);
      expect(x.approximately(Angle.degrees(44.9), 0.1), true);

      expect(x.approximately(Angle.degrees(45.1), 0.01), true);
      expect(x.approximately(Angle.degrees(44.9), 0.01), true);
    });
    test("Normalization", () {
      expect(Angle.turns(0.0).normalized, Angle.turns(0.0));
      expect(Angle.turns(-0.0).normalized, Angle.turns(0.0));
      expect(Angle.turns(1.0).normalized, Angle.turns(0.0));

      expect(Angle.turns(0.3).normalized, Angle.turns(0.3));
      expect(Angle.turns(2.5).normalized, Angle.turns(0.5));

      expect(Angle.turns(-0.25).normalized, Angle.turns(0.75));
      expect(Angle.turns(-2.25).normalized, Angle.turns(0.75));
    });
  });
  group("Trigonometric functions", () {
    test("Sinusoidals", () {
      final Angle x = Angle.degrees(45);
      final double epsilon = 1e-9;
      expect(Angle.asin(x.sin).approximately(x, epsilon), true);
      expect(Angle.acos(x.cos).approximately(x, epsilon), true);
      expect(Angle.atan(x.tan).approximately(x, epsilon), true);

      expect(Angle.atan2(0, 0), Angle.degrees(0));
      expect(Angle.atan2(0, 1), Angle.degrees(0));
      expect(Angle.atan2(1, 1), Angle.degrees(45));
      expect(Angle.atan2(1, 0), Angle.degrees(90));
      expect(Angle.atan2(1, -1), Angle.degrees(135));
      expect(Angle.atan2(0, -1), Angle.degrees(180));
      expect(Angle.atan2(-1, -1), Angle.degrees(-135));
      expect(Angle.atan2(-1, 0), Angle.degrees(-90));
      expect(Angle.atan2(-1, 1), Angle.degrees(-45));

      expect(Angle.atanFullTurn(0, 0), Angle.degrees(0));
      expect(Angle.atanFullTurn(0, 1), Angle.degrees(0));
      expect(Angle.atanFullTurn(1, 1), Angle.degrees(45));
      expect(Angle.atanFullTurn(1, 0), Angle.degrees(90));
      expect(Angle.atanFullTurn(1, -1), Angle.degrees(135));
      expect(Angle.atanFullTurn(0, -1), Angle.degrees(180));
      expect(Angle.atanFullTurn(-1, -1), Angle.degrees(225));
      expect(Angle.atanFullTurn(-1, 0), Angle.degrees(270));
      expect(Angle.atanFullTurn(-1, 1), Angle.degrees(315));

      expect(Angle.cartesian(0, 0), Angle.degrees(0));
      expect(Angle.cartesian(1, 0), Angle.degrees(0));
      expect(Angle.cartesian(1, 1), Angle.degrees(45));
      expect(Angle.cartesian(0, 1), Angle.degrees(90));
      expect(Angle.cartesian(-1, 1), Angle.degrees(135));
      expect(Angle.cartesian(-1, 0), Angle.degrees(180));
      expect(Angle.cartesian(-1, -1), Angle.degrees(225));
      expect(Angle.cartesian(0, -1), Angle.degrees(270));
      expect(Angle.cartesian(1, -1), Angle.degrees(315));
    });
  });
  group("Compatibility", () {
    test("Constructors", () {
      // ignore: deprecated_member_use_from_same_package
      expect(Angle.fromDegrees(45), Angle.degrees(45));
      // ignore: deprecated_member_use_from_same_package
      expect(Angle.fromGradians(45), Angle.gradians(45));
      // ignore: deprecated_member_use_from_same_package
      expect(Angle.fromRadians(45), Angle.radians(45));
      // ignore: deprecated_member_use_from_same_package
      expect(Angle.fromTurns(45), Angle.turns(45));
    });
  });
}


lib/src/angles_base.dart
import 'dart:math' as math;

import 'package:meta/meta.dart';

/// Represents an angle.
/// The unit used for construction does not matter, as angles internally
/// always use radians.
/// The class is designed to be immutable but provides operators like adding
/// or subtracting, which all create new angles instead of mutating existing
/// ones.
@immutable
class Angle implements Comparable<Angle> {
  final double _storage;

  /// Create a zero angle.
  const Angle.zero() : _storage = 0.0;

  /// Create an angle defined by degrees.
  /// One full turn equals 360 degrees.
  const Angle.fromDegrees(final double degrees)
      : _storage = degrees / 180.0 * math.pi;

  /// Create an angle from radians.
  /// One full turn equals 2pi radians.
  const Angle.fromRadians(final double radians) : _storage = radians;

  /// Create an angle defined by gradians.
  /// One full turn equals 400 gradians.
  const Angle.fromGradians(final double gradians)
      : _storage = gradians / 200.0 * math.pi;

  /// Create an angle defined by turns.
  /// One full turn equals 1 turn.
  const Angle.fromTurns(final double turns) : _storage = turns * 2.0 * math.pi;

  /// Create an angle defining one full turn.
  factory Angle.fullTurn() => Angle.fromTurns(1.0);

  /// Create an angle defining one half turn.
  factory Angle.halfTurn() => Angle.fromTurns(0.5);

  /// Create an angle defining one quarter turn.
  factory Angle.quarterTurn() => Angle.fromTurns(0.25);

  /// Create an angle defining one third turn.
  factory Angle.thirdTurn() => Angle.fromTurns(1.0 / 3.0);

  /// Create an angle defining one eighth turn.
  factory Angle.eighthTurn() => Angle.fromTurns(0.125);

  /// Create an angle by computing the arc sine of [c].
  Angle.asin(final double c) : _storage = math.asin(c);

  /// Create an angle by computing the arc cosine of [c].
  Angle.acos(final double c) : _storage = math.acos(c);

  /// Create an angle by computing the arc tangent of [c].
  Angle.atan(final double c) : _storage = math.atan(c);

  /// Create an angle by computing the arc tangent from `y / x`.
  /// This uses the `atan2()` function, therefore there are two
  /// separate parameters for the nominator and the denominator.
  /// The result is an angle in `]-180°;180°]`.
  /// Sample values:
  ///  - `y=0, x=0` => 0°
  ///  - `y=0, x=1` => 0°
  ///  - `y=1, x=0` => 90°
  ///  - `y=0, x=-1` => 180°
  ///  - `y=-1, x=0` => -90°
  Angle.atan2(final double y, final double x) : _storage = math.atan2(y, x);

  /// Works in the same manner as [Angle.atan2], but results
  /// in an angle in `]0.0°;360°]`.
  /// Sample values:
  ///  - `y=0, x=0` => 0°
  ///  - `y=0, x=1` => 0°
  ///  - `y=1, x=0` => 90°
  ///  - `y=0, x=-1` => 180°
  ///  - `y=-1, x=0` => 270°
  factory Angle.atanFullTurn(final double y, final double x) {
    if (y >= 0) {
      return Angle.atan2(y, x);
    } else {
      return Angle.atan2(-y, -x) + Angle.halfTurn();
    }
  }

  /// Get the angle in turns.
  double get turns => (_storage) / math.pi / 2.0;

  /// Get the angle in degrees.
  double get degrees => (_storage / math.pi) * 180.0;

  /// Get the angle in radians.
  double get radians => _storage;

  /// Get the angle in gradians.
  double get gradians => (_storage / math.pi) * 200.0;

  /// Compute the sine of this angle.
  double get sin => math.sin(radians);

  /// Compute the cosine of this angle.
  double get cos => math.cos(radians);

  /// Compute the tangent of this angle.
  double get tan => math.tan(radians);

  /// Accumulate this and [other], returning a new angle.
  Angle operator +(final Angle other) =>
      Angle.fromRadians(radians + other.radians);

  /// Compute the difference of this to [other], returning a new angle.
  Angle operator -(final Angle other) =>
      Angle.fromRadians(radians - other.radians);

  /// Negate this angle, returning a new angle.
  Angle operator -() => Angle.fromRadians(-radians);

  /// Scale this angle, returning a new angle.
  Angle operator *(final double scale) => Angle.fromRadians(radians * scale);

  /// Inverse scale this angle, returning a new angle.
  Angle operator /(final double scale) => Angle.fromRadians(radians / scale);

  /// Return a string representation of this angle, optimized to be readable.
  /// The angle is printed in degrees, and only one decimal digit is included.
  @override
  String toString() => "${degrees.toStringAsFixed(1)}°";

  @override
  int compareTo(Angle other) {
    if (this == other) {
      return 0;
    } else if (this > other) {
      return 1;
    } else {
      return -1;
    }
  }

  @override
  int get hashCode => _storage.hashCode;

  /// Compares this to [other].
  bool operator <(final Angle other) => _storage < other._storage;

  /// Compares this to [other].
  bool operator >(final Angle other) => _storage > other._storage;

  /// Compares this to [other].
  bool operator <=(final Angle other) => _storage <= other._storage;

  /// Compares this to [other].
  bool operator >=(final Angle other) => _storage >= other._storage;

  /// Compares to another angle.
  /// Since IEEE 754 floating points are approximations to real numbers, they cannot
  /// be really tested for equality.
  @override
  bool operator ==(covariant final Angle rhs) => _storage == rhs._storage;

  /// Checks if [rhs] is approximately this angle, +/- [range].
  /// [range] must be non-negative.
  bool approximately(final Angle rhs, final double range) {
    assert(range >= 0);
    return rhs._storage >= _storage - range && rhs._storage <= _storage + range;
  }
}


lib/angles.dart
library angles;

import 'dart:math' as math;

import 'package:meta/meta.dart';

/// Represents an angle.
/// The unit used for construction does not matter, as angles internally
/// always use radians.
/// The class is designed to be immutable but provides operators like adding
/// or subtracting, which all create new angles instead of mutating existing
/// ones.
@immutable
class Angle implements Comparable<Angle> {
  final double _storage;

  /// Create a zero angle.
  const Angle.zero() : _storage = 0.0;

  /// Create an angle defined by degrees.
  /// One full turn equals 360 degrees.
  const Angle.degrees(final double degrees)
      : _storage = degrees / 180.0 * math.pi;

  @Deprecated("Use Angle.degrees instead")
  const Angle.fromDegrees(final double degrees)
      : _storage = degrees / 180.0 * math.pi;

  /// Create an angle from radians.
  /// One full turn equals 2pi radians.
  const Angle.radians(final double radians) : _storage = radians;

  @Deprecated("Use Angle.radians instead")
  const Angle.fromRadians(final double radians) : _storage = radians;

  /// Create an angle defined by gradians.
  /// One full turn equals 400 gradians.
  const Angle.gradians(final double gradians)
      : _storage = gradians / 200.0 * math.pi;

  @Deprecated("Use Angle.gradians instead")
  const Angle.fromGradians(final double gradians)
      : _storage = gradians / 200.0 * math.pi;

  /// Create an angle defined by turns.
  /// One full turn equals 1 turn.
  const Angle.turns(final double turns) : _storage = turns * 2.0 * math.pi;

  @Deprecated("Use Angle.turns instead")
  const Angle.fromTurns(final double turns) : _storage = turns * 2.0 * math.pi;

  /// Create an angle defining one full turn.
  factory Angle.fullTurn() => Angle.turns(1.0);

  /// Create an angle defining one half turn.
  factory Angle.halfTurn() => Angle.turns(0.5);

  /// Create an angle defining one quarter turn.
  factory Angle.quarterTurn() => Angle.turns(0.25);

  /// Create an angle defining one third turn.
  factory Angle.thirdTurn() => Angle.turns(1.0 / 3.0);

  /// Create an angle defining one eighth turn.
  factory Angle.eighthTurn() => Angle.turns(0.125);

  /// Create an angle by computing the arc sine of [c].
  Angle.asin(final double c) : _storage = math.asin(c);

  /// Create an angle by computing the arc cosine of [c].
  Angle.acos(final double c) : _storage = math.acos(c);

  /// Create an angle by computing the arc tangent of [c].
  Angle.atan(final double c) : _storage = math.atan(c);

  /// Create an angle by computing the arc tangent from `y / x`.
  /// This uses the `atan2()` function, therefore there are two
  /// separate parameters for the nominator and the denominator.
  /// The result is an angle in `]-180°;180°]`.
  /// Sample values:
  ///  - `y=0, x=0` => 0°
  ///  - `y=0, x=1` => 0°
  ///  - `y=1, x=0` => 90°
  ///  - `y=0, x=-1` => 180°
  ///  - `y=-1, x=0` => -90°
  Angle.atan2(final double y, final double x) : _storage = math.atan2(y, x);

  /// Works in the same manner as [Angle.atan2], but results
  /// in an angle in `[0°, 360°)`.
  /// Sample values:
  ///  - `y=0, x=0` => 0°
  ///  - `y=0, x=1` => 0°
  ///  - `y=1, x=0` => 90°
  ///  - `y=0, x=-1` => 180°
  ///  - `y=-1, x=0` => 270°
  factory Angle.atanFullTurn(final double y, final double x) {
    if (y >= 0) {
      return Angle.atan2(y, x);
    } else {
      return Angle.atan2(-y, -x) + Angle.halfTurn();
    }
  }

  /// Returns an angle in `[0°, 360°)`, pointing to the cartesian coordinate.
  /// The `x` axis has angle `0°`, the `y` axis has angle `90°`.
  factory Angle.cartesian(final double x, final double y) {
    return Angle.atanFullTurn(y, x);
  }

  /// Get the angle in turns.
  double get turns => (_storage) / math.pi / 2.0;

  /// Get the angle in degrees.
  double get degrees => (_storage / math.pi) * 180.0;

  /// Get the angle in radians.
  double get radians => _storage;

  /// Get the angle in gradians.
  double get gradians => (_storage / math.pi) * 200.0;

  /// Compute the sine of this angle.
  double get sin => math.sin(radians);

  /// Compute the cosine of this angle.
  double get cos => math.cos(radians);

  /// Compute the tangent of this angle.
  double get tan => math.tan(radians);

  /// Accumulate this and [other], returning a new angle.
  Angle operator +(final Angle other) => Angle.radians(radians + other.radians);

  /// Compute the difference of this to [other], returning a new angle.
  Angle operator -(final Angle other) => Angle.radians(radians - other.radians);

  /// Negate this angle, returning a new angle.
  Angle operator -() => Angle.radians(-radians);

  /// Scale this angle, returning a new angle.
  Angle operator *(final double scale) => Angle.radians(radians * scale);

  /// Inverse scale this angle, returning a new angle.
  Angle operator /(final double scale) => Angle.radians(radians / scale);

  /// Returns the ratio between two angles.
  double ratio(final Angle other) => radians / other.radians;

  /// Return a string representation of this angle, optimized to be readable.
  /// The angle is printed in degrees, and only one decimal digit is included.
  @override
  String toString() => "${degrees.toStringAsFixed(1)}°";

  /// Returns an angle with the some direction, but confined to
  /// the interval `[0°, 360°)`.
  Angle get normalized {
    if (_storage == 0) {
      return Angle.zero();
    } else if (_storage > 0) {
      return Angle.turns(turns.remainder(1.0));
    } else {
      return Angle.fullTurn() + Angle.turns(turns.remainder(1.0));
    }
  }

  // Compare to another angle. Returns:
  // - `0`: Both angles are equal.
  // - `1`: This angle is greater.
  // - `-1`: This angle is smaller.
  @override
  int compareTo(Angle other) {
    if (this == other) {
      return 0;
    } else if (this > other) {
      return 1;
    } else {
      return -1;
    }
  }

  @override
  int get hashCode => _storage.hashCode;

  /// Compares this to [other].
  bool operator <(final Angle other) => _storage < other._storage;

  /// Compares this to [other].
  bool operator >(final Angle other) => _storage > other._storage;

  /// Compares this to [other].
  bool operator <=(final Angle other) => _storage <= other._storage;

  /// Compares this to [other].
  bool operator >=(final Angle other) => _storage >= other._storage;

  /// Compares to another angle.
  /// Since IEEE 754 floating points are approximations to real numbers, they cannot
  /// be really tested for equality.
  @override
  bool operator ==(covariant final Angle rhs) => _storage == rhs._storage;

  /// Checks if [rhs] is approximately this angle, +/- [range].
  /// [range] must be non-negative.
  bool approximately(final Angle rhs, final double range) {
    assert(range >= 0);
    return rhs._storage >= _storage - range && rhs._storage <= _storage + range;
  }
}
